Introduce MOAB coupler interfaces and update moab driver

rljacob opened the pull request:
Add interfaces to driver-moab in WaterCycle components:  EAM, ELM, MOSART, MPAS-Ocean, MPAS-SeaIce.  Summary of changes is below.

MOAB interfaces and support routines in components are isolated with a "HAVE_MOAB" ifdef that will be automatically defined if the MOAB coupler is chosen.  This allows components to use either mct or moab couplers for now.  Choice is made during create_newcase.  There are also cpp defines for debugging.  MOABDEBUG will output .h5m files throughout the coupling processs and additional log info (very slow).  MOABCOMP will calculate diffs between MCT and MOAB coupling fields.  These defines need to be set by hand if you wish to use them.  These will be removed later.    When using default driver-mct, there is no impact on answers and no dependency on MOAB.

Also bring in latest version of driver-moab which is feature-complete for the watercycle case.  All maps (except r2o) are calculated online.  Merging, fluxes and all other coupler calculations are performed with moab data structures.  Currently includes MCT-scaffolding for verification (data moving between components follows both MCT- and MOAB- paths; offline maps still read by MCT; all mct-based datatypes still present).

[BFB]

### Build system mods
- If USE_MOAB is defined, call find_package(MOAB), add MOAB include dirs and HAVE_MOAB cpp.  
- Also set MOAB_ROOT on chrysalis and gce.
- See also mods to MPAS build


### EAM mods
- cpl/atm_comp_mct.F90 - define and call routines to set the physics grid "cloud" (used when physics is on the SE points) and also define routines to export/import EAM data to/from MOAB meshes and call them when HAVE_MOAB is defined.
- dynamics/se/semoab_mod.F90 - new module to hold routines for defining SE meshes (coarse, fine and physgrid) to MOAB
- dynamics/se/dyn_comp.F90 - define MOAB apps for each mesh and call create_moab_meshes
- eam/bld/config_files/definition.xml - add moab as interface option
- eam/bld/configure - add moab as comp_intf option
- eam/cime_config/buildnml - add moab as interface option

 ### ELM MODS
- cpl/lnd_comp_mct.F90: define and call the routine to define the ELM MOAB app and describe the mesh.  Also define and use routines to export/import data to/from the MOAB coupler.
- main/srfrdMod.F90:  calculate vertex locations for MOAB from domain file info.
- config_definitions.xml: add moab as comp_intf option

### MOSART MODS
- bld/build-namelist and namelist_default* files:  define and set new namelist var for mesh frivinp_mesh
- cpl/rof_comp_mct.F90:  define and call the routine to define the MOSART MOAB app and describe the mesh.  Also define and use routines to export/import data to/from the MOAB coupler.
- riverroute/RtmMod.F90, RtmVar.F90:  read in the new frivinp_mesh file.  Pass to coupler infobuffer.

### MPAS MODS
- mpas-framework/Makefile:  Add HAVE_MOAB and MOAB include paths
- mpas-framework/src/Makefile, CMakelists.txt:  Add MOAB parts if COMP_INTERFACE is moab
- mpas-framework/driver/mpas-subdriver.F:  init MOAB if using this driver
- mpas-framework/framework/mpas-moabmesh.F:  new routine to define MPAS mesh to MOAB.  Works for all MPAS components
- mpas-ocean/driver/ocn_comp_mct.F and
- mpas-seaice/driver/ice_comp_mct.F:   call routine to define mesh to MOAB.   Define and call routines to export/import data to/from MOAB coupler

rljacob commented:
@jonbob @bishtgautam @wlin7 @liho745 : you only need to look at the mods to components.  @jgfouca Just look at build system mods.

jgfouca reviewed (changes requested):
I think you can get rid of the MOAB_PATH stuff in the cmake_macros. You are setting MOAB_ROOT in the config_machines.xml environment, which should be sufficient to activate and find MOAB.

jgfouca commented on components/cmake/build_model.cmake:
```diff
@@ -243,6 +247,11 @@ function(build_model COMP_CLASS COMP_NAME)
+    if (USE_MOAB)
+      target_link_libraries(${TARGET_NAME} PRIVATE ${MOAB_LIBRARIES})
+      target_include_directories(${TARGET_NAME} PRIVATE ${MOAB_INCLUDE_DIRS})
+      message("JGF adding include directories ${MOAB_INCLUDE_DIRS}")
```
You can get rid of my debugging print.

jgfouca commented on components/cmake/build_model.cmake:
```diff
@@ -233,6 +233,10 @@ function(build_model COMP_CLASS COMP_NAME)
     endforeach()
+    if (USE_MOAB)
+      target_include_directories(${TARGET_NAME} PRIVATE ${MOAB_INCLUDE_DIRS})
+      message("JGF adding include directories ${MOAB_INCLUDE_DIRS}")
```
You can get rid of my debugging print.

bishtgautam commented:
@rljacob An overview question: Why isn't the big `#ifdef HAVE_MOAB` in [components/elm/src/cpl/lnd_comp_mct.F90](https://github.com/E3SM-Project/E3SM/pull/5979/files#diff-dcaa702776bbd34ce3a03cb4e3a731f944062eced731f6106b30a101ca949c21R834-R1645) put in a new file `components/elm/src/cpl/lnd_comp_moab.F90`?

rljacob commented:
Because we're still running the mct and moab parts at the same time.   The separation will be done later.

rljacob commented:
Forgot to mention how to run this.  On Chrysalis, use "--res ne4pg2_oQU480 --compset WCYCL1850NS --driver moab" in your create_newcase command.

rljacob commented:
@jgfouca the particular version of master this started from still needs MOAB_PATH in the cmake macros or there's a build error early:  
```
/gpfs/fs1/home/jacob/E3SMmb/cime/scripts/Wne4pgmb4/Tools/Makefile:244: *** MOAB_PATH must be defined when using moab driver.  Stop.
ERROR: cat: Filepath: No such file or directory
```
I'll remove those in a follow-up PR.

jgfouca commented:
@rljacob, why not do an upstream merge to get onto latest master? It looks like there are conflicts to resolve so you'll have to do this anyway.

rljacob commented:
I prefer to resolve conflicts during the merge unless there's a bunch of them.  I already merged master to this branch once for some build things.

rljacob commented:
@jgfouca I did a test-merge to next and the conflicts are trivial.   Also I removed those debugging prints from cmake.

rljacob commented:
@jonbob @bishtgautam @liho745 please review.

jonbob commented on components/mpas-ocean/driver/ocn_comp_mct.F:
@rljacob -- it doesn't make more sense to have separate ocn_comp_mct and ocn_comp_moab files?

rljacob replied:
Gautum had the same question about lnd_comp_mct.  We're keeping one source while the MCT-MOAB verification code is in place.  They'll be separated when that's removed.

bishtgautam reviewed (approved):
Land changes look good.

liho745 reviewed (approved):
River-related changes look good.

jonbob reviewed (approved):
Looks good from a visual inspection

rljacob commented:
@wlin7 if you want any changes, let me know before Wednesday.  I'm going to merge it then.

rljacob commented:
merged to next

rljacob commented:
For the record here are the files that merged cleanly ("Changes to be committed"), files that had to have conflicts fixed ("Unmerged paths"), and files that needed further modification ("Changes not staged") to work with the new build system.

```
Changes to be committed:
	modified:   components/CMakeLists.txt
	modified:   components/cice/bld/config_files/definition.xml
	modified:   components/cmake/find_dep_packages.cmake
	modified:   components/eam/bld/config_files/definition.xml
	modified:   components/eam/bld/configure
	modified:   components/eam/cime_config/buildnml
	modified:   components/eam/src/cpl/atm_comp_mct.F90
	modified:   components/eam/src/dynamics/se/dyn_comp.F90
	new file:   components/eam/src/dynamics/se/semoab_mod.F90
	modified:   components/elm/bld/config_files/config_definition.xml
	modified:   components/elm/src/cpl/lnd_comp_mct.F90
	modified:   components/elm/src/main/surfrdMod.F90
	modified:   components/elm/src/utils/domainMod.F90
	modified:   components/mosart/bld/build-namelist
	modified:   components/mosart/bld/namelist_files/namelist_defaults_mosart.xml
	modified:   components/mosart/bld/namelist_files/namelist_definition_mosart.xml
	modified:   components/mosart/cime_config/testdefs/testmods_dirs/mosart/mos_usrdat/user_nl_mosart
	modified:   components/mosart/cime_config/testdefs/testmods_dirs/mosart/sediment/shell_commands
	modified:   components/mosart/cime_config/testdefs/testmods_dirs/mosart/sediment/user_nl_mosart
	modified:   components/mosart/cime_config/testdefs/testmods_dirs/mosart/unstructure/user_nl_mosart
	modified:   components/mosart/src/cpl/rof_comp_mct.F90
	modified:   components/mosart/src/riverroute/RtmMod.F90
	modified:   components/mosart/src/riverroute/RtmVar.F90
	modified:   components/mpas-framework/Makefile
	modified:   components/mpas-framework/src/CMakeLists.txt
	modified:   components/mpas-framework/src/Makefile
	modified:   components/mpas-framework/src/driver/mpas_subdriver.F
	modified:   components/mpas-framework/src/framework/Makefile
	new file:   components/mpas-framework/src/framework/mpas_moabmesh.F
	modified:   components/mpas-ocean/driver/ocn_comp_mct.F
	modified:   driver-mct/cime_config/config_component.xml
	modified:   driver-moab/cime_config/buildexe
	new file:   driver-moab/cime_config/buildlib_cmake
	modified:   driver-moab/cime_config/buildnml
	modified:   driver-moab/cime_config/config_component.xml
	modified:   driver-moab/cime_config/config_component_cesm.xml
	modified:   driver-moab/cime_config/config_component_e3sm.xml
	modified:   driver-moab/cime_config/config_pes.xml
	modified:   driver-moab/cime_config/namelist_definition_drv.xml
	modified:   driver-moab/cime_config/namelist_definition_drv_flds.xml
	modified:   driver-moab/cime_config/namelist_definition_modelio.xml
	modified:   driver-moab/main/cime_comp_mod.F90
	modified:   driver-moab/main/cime_driver.F90
	modified:   driver-moab/main/component_mod.F90
	modified:   driver-moab/main/component_type_mod.F90
	modified:   driver-moab/main/cplcomp_exchange_mod.F90
	modified:   driver-moab/main/prep_aoflux_mod.F90
	modified:   driver-moab/main/prep_atm_mod.F90
	new file:   driver-moab/main/prep_iac_mod.F90
	modified:   driver-moab/main/prep_ice_mod.F90
	modified:   driver-moab/main/prep_lnd_mod.F90
	modified:   driver-moab/main/prep_ocn_mod.F90
	modified:   driver-moab/main/prep_rof_mod.F90
	new file:   driver-moab/main/seq_diagBGC_mct.F90
	modified:   driver-moab/main/seq_diag_mct.F90
	modified:   driver-moab/main/seq_flux_mct.F90
	modified:   driver-moab/main/seq_frac_mct.F90
	modified:   driver-moab/main/seq_hist_mod.F90
	modified:   driver-moab/main/seq_io_mod.F90
	modified:   driver-moab/main/seq_map_mod.F90
	modified:   driver-moab/main/seq_map_type_mod.F90
	modified:   driver-moab/main/seq_rest_mod.F90
	modified:   driver-moab/shr/seq_comm_mct.F90
	modified:   driver-moab/shr/seq_drydep_mod.F90
	modified:   driver-moab/shr/seq_flds_mod.F90
	modified:   driver-moab/shr/seq_infodata_mod.F90
	new file:   driver-moab/shr/seq_pauseresume_mod.F90
	new file:   driver-moab/shr/shr_dust_mod.F90
	new file:   driver-moab/shr/shr_fan_mod.F90

Unmerged paths:
  (use "git add <file>..." to mark resolution)
	both modified:   cime_config/machines/cmake_macros/gnu_anlgce.cmake
	both modified:   cime_config/machines/cmake_macros/gnu_chrysalis.cmake
	both modified:   cime_config/machines/cmake_macros/intel_chrysalis.cmake
	both modified:   cime_config/machines/config_machines.xml
	both modified:   components/cmake/build_model.cmake
	both modified:   components/mpas-ocean/driver/mpaso_cpl_indices.F
	both modified:   components/mpas-seaice/driver/ice_comp_mct.F

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
  (commit or discard the untracked or modified content in submodules)
	modified:   components/cmake/modules/FindCsmShare.cmake
	modified:   share/CMakeLists.txt
```

jgfouca commented:
Congrats! This was a big change

jgfouca commented:
Do you want me to take a second look at any of those cmake conflicts?

rljacob commented:
Thanks!  You don't have to look at anything.  You had already shown how to fix it in  https://github.com/rljacob/E3SM/tree/jgfouca/moab_fixes.

What I didn't try was adding those changes to my branch so I wouldn't have to do it during the merge.  This branch only had some of your build changes so wasn't sure it would work.
